### PR TITLE
93 add database interface

### PR DESCRIPTION
**Pull request by AlSchlo**
Add a fancy database interface and its implementation with Firebase Realtime Database, according to #93. This contains:

* Sign in tests adapted to work with preferences (cf previous sprint)
* A generic database interface
* A Firebase implementation
* An enumeration containing all databases useful for H3LP
* Demo code to show basic communication (help call)
* Firebase dependent tests to increase coverage

-> Here is a nice meme:
image

### CODE REVIEW

File: app/src/main/java/com/github/h3lp3rs/h3lp/database/FireDatabase.kt [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        }
+        openListeners.remove(key)
+    }
+
+    /**
+     * Deletes an entry of a given key from the database
+     * @param key They key in the database
+     */
+    override fun delete(key: String) {
+        db.child(key).removeValue()
```
Comment (Ahayderi) [inline app/src/main/java/com/github/h3lp3rs/h3lp/database/FireDatabase.kt:159]: I am not at all sure about this but shouldn't deleting a key also delete its listeners in openListeners by removing the key (I guess that the respective listeners on the actual database are deleted if you delete a key so there could be an inconsistency)?

**Review by Ahayderi — APPROVED** [review]
This was a great implementation, the code is neat and well documented! I also found the Database abstraction to be very clear and user-friendly so it will be a delight to use in the coming weeks.

File: app/src/androidTest/java/com/github/h3lp3rs/h3lp/FirebaseTest.kt [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            db.setInt(testingKey + 1, 1)
+            db.setString(testingKey + 2, "1")
+            db.setBoolean(testingKey + 3, false)
+            db.setDouble(testingKey + 4, 1.1)
+            db.setObject(testingKey + 5, Foo::class.java, foo)
+            assertEquals(1, db.getInt(testingKey + 1).get())
+            assertEquals("1", db.getString(testingKey + 2).get())
+            assertEquals(false, db.getBoolean(testingKey + 3).get())
+            assertEquals(1.1, db.getDouble(testingKey + 4).get())
+            assertEquals(foo, db.getObject(testingKey + 5, Foo::class.java).get())
```
[the remark this answers is not in the document]
Comment (AlSchlo) [reply app/src/androidTest/java/com/github/h3lp3rs/h3lp/FirebaseTest.kt:35]: Nasty (failed) temporary tricks to boost the coverage :p

File: app/src/main/java/com/github/h3lp3rs/h3lp/database/FireDatabase.kt (hunk 1 above)
Comment (AlSchlo) [reply app/src/main/java/com/github/h3lp3rs/h3lp/database/FireDatabase.kt:159]: Nice catch, we implemented deletion before the listeners, so we just forgot it

File: app/src/main/java/com/github/h3lp3rs/h3lp/MedicalCardActivity.kt [hunk 3]
```diff
@@ -182,7 +182,7 @@ class MedicalCardActivity : AppCompatActivity() {
             override fun onClick(widget: View) {
                 //This avoid checkbox checking when clicking on the policy link
                 //Sadly it is incompatible with Expresso onClick() so is commented for cirrus
```
Comment (wiam8) [inline app/src/main/java/com/github/h3lp3rs/h3lp/MedicalCardActivity.kt:184]:
```suggestion
                // This avoids checkbox checking when clicking on the policy link
                // Sadly it is incompatible with Expresso onClick(), so it is commented for cirrus
```

File: app/src/main/java/com/github/h3lp3rs/h3lp/database/FireDatabase.kt [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+import com.google.firebase.database.DatabaseError
+import com.google.firebase.database.DatabaseReference
+import com.google.firebase.database.ValueEventListener
+import com.google.firebase.database.ktx.*
+import com.google.firebase.ktx.Firebase
+import com.google.gson.Gson
+import java.util.concurrent.CompletableFuture
+
+/**
+ * Implementation of an NoSQL external database based on Firebase
```
Comment (wiam8) [inline app/src/main/java/com/github/h3lp3rs/h3lp/database/FireDatabase.kt:13]:
```suggestion
 * Implementation of a NoSQL external database based on Firebase

```

File: app/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -113,6 +113,11 @@
     <string name="condition_help_msg">This Field is dedicated to any illness or diseases that a caregiver should know in case of emergency</string>
     <string name="treatment_help_msg">This Field is dedicated to any medication your are taking that a caregiver should know in case of emergency</string>
     <string name="allergy_help_msg">This Field is dedicated to any allergy you suffer from that a caregiver should know in case of emergency</string>
+    <string name="ventolin_db_key"> VENTOLIN</string>
+    <string name="help">H3LP!</string>
+    <string name="nothing">nothing</string>
+    <string name="emergency">Emergency</string>
+    <string name="need_help"> Hind need a ventolin </string>
```
Comment (wiam8) [inline app/src/main/res/values/strings.xml:120]:
```suggestion
    <string name="need_help"> Hind needs a ventolin </string>
```

**Review by wiam8 — APPROVED** [review]
Great work! Very developer-friendly, we won't have any trouble using it in the coming weeks.

**Pull request merged**